test/test_cpp_api_parity.py

```diff
@@ -0,0 +1,179 @@
+            test_params = process_test_params(test_params, python_module_class, cpp_namespace)
+            module_name = test_params['module_name']
+            if module_name not in test_params_map:
+                test_params_map[module_name] = []
+            test_params_map[module_name].append(test_params)
+
+        # Generate C++ code for each test case
+        cpp_source = TORCH_NN_MODULE_COMMON_TEST_HARNESS + test_params['cpp_source']
+        functions = []
+        for module_name in module_names:
+            for test_params in test_params_map[module_name]:
+                cpp_source += TORCH_NN_MODULE_WRAPPER.substitute(
+                    module_variant_name=test_params['module_variant_name'],
+                    module_qualified_name=test_params['cpp_namespace'] + test_params['module_name'],
+                    cpp_constructor_args=test_params['cpp_constructor_args'])
+                for method in torch_nn_test_methods:
+                    functions.append(test_params['module_variant_name'] + '_test_' + method)
+
+        # Just-in-time compile the C++ test code
+        cpp_module = torch.utils.cpp_extension.load_inline(
+            name=test_suite_name,
+            cpp_sources=cpp_source,
+            functions=functions,
+            verbose=True,
+        )
+
+        def test_method(method_name, test_params):
+            expect_error = test_params['expect_error']
+            input_size = test_params['input_size']
+            python_module_class = test_params['python_module_class']
+            python_constructor_args = test_params['python_constructor_args']
+            module_variant_name = test_params['module_variant_name']
+            test_name = module_variant_name + '_test_' + method_name
+            example_input = torch.randn(input_size)
+            with common.freeze_rng_state():
+                torch.manual_seed(2)
+                module = python_module_class(*python_constructor_args)
+                if method_name == 'forward':
+                    python_output = module(example_input)
+                # We use JIT tracing to transfer Python module state to C++
+                traced_script_module = torch.jit.trace(module, example_input)
+                # Ideally we would like to not have to manually delete the file, but NamedTemporaryFile
+                # opens the file, and it cannot be opened multiple times in Windows. To support Windows,
+                # close the file after creation and try to remove it manually
+                f = tempfile.NamedTemporaryFile(delete=False)
+                try:
+                    f.close()
+                    traced_script_module.save(f.name)
+                    torch.manual_seed(2)
+                    if method_name == 'init':
+                        args = (f.name, )
+                    elif method_name == 'forward':
+                        args = (f.name, example_input, python_output)
+                    if expect_error:
```

Comment:
This sounds like it should be an [expectedFailure](https://docs.python.org/3/library/unittest.html#unittest.expectedFailure)